flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/FlinkTypeFactory.scala

```diff
@@ -293,6 +293,11 @@ class FlinkTypeFactory(typeSystem: RelDataTypeSystem) extends JavaTypeFactoryImp
       case it: TimeIndicatorRelDataType =>
         new TimeIndicatorRelDataType(it.typeSystem, it.originalType, isNullable, it.isEventTime)
 
+      // for nested rows we keep the nullability property,
+      // top-level rows fall back to Calcite's default handling
+      case rt: RelRecordType if rt.getStructKind == StructKind.PEEK_FIELDS_NO_EXPAND =>
```

Comment:
I find that the comment in `RelDataTypeFactoryImpl#createTypeWithNullability` says `According to the SQL standard, nullability for struct types can be defined only for columns, which translates to top level structs. Nested struct attributes are always nullable, so in principle we could always set the nested attributes to be nullable.` 
we do not follow the SQL standard here ?

Reply:
Yes, I also saw this comment. That's why I only fixed a bug around nested rows in this PR. However, if we are looking for more efficient plans we should further investigate this shortcoming. I recommend to remove `if rt.getStructKind == StructKind.PEEK_FIELDS_NO_EXPAND` to see the resulting plans esp. around joins. Would definitely be a follow-up issue.

Reply:
Make sense to me. I also notice `outer joins`  is mentioned in `RelDataTypeFactoryImpl#copyRecordType`'s  comment